Erlang Sampler and Resource documentation (#2556)
open-telemetry/opentelemetry.io: content/en/docs/instrumentation/erlang/sampling.md

---
title: Sampling
weight: 37
---

> 
> ## Sampling

Reviewer: ```suggestion
```

[Sampling](/docs/concepts/sampling/) is a process that restricts the amount of
traces that are generated by a system. The Erlang SDK offers several
[head samplers](/docs/concepts/sampling#head-sampling).

> ### Default behavior

Reviewer: ```suggestion
## Default behavior
```

By default, all spans are sampled, and thus, 100% of traces are sampled. If you

> do not need to manage data volume, don't bother setting a sampler.

Reviewer 2: I am not sure if this is the right language.

```suggestion
do not need to manage data volume, you do not need to configure a sampler.
```

> ### ParentBasedSampler

Reviewer: ```suggestion
## ParentBasedSampler
```

When sampling, the `ParentBasedSampler` is most often used for

> [head sampler](/docs/concepts/sampling/#head-sampling). It uses the sampling

Reviewer 3: ```suggestion
[head sampling](/docs/concepts/sampling/#head-sampling). It uses the sampling
```
?

decision of the Span's parent, or the fact that there is no parent, to know
which secondary sampler to use.

> The sampler can be configured with environment variables `OTEL_TRACES_SAMPLER`

Reviewer 3: ```suggestion
The sampler can be configured with the environment variables `OTEL_TRACES_SAMPLER`
```

and `OTEL_TRACES_SAMPLER_ARG` or using the Application configuration allows you
to configure each of the 5 potential states of a Span's parent:

- `root` - No parent
- `remote_parent_sampled` - Parent is from a remote Span that is sampled
- `remote_parent_not_sampled` - Parent is from a remote Span that is not sampled
- `local_parent_sampled` - Parent is from a local Span that is sampled
- `local_parent_not_sampled` - Parent is from a local Span that is not sampled

> #### TraceIdRatioBasedSampler

Reviewer: ```suggestion
### TraceIdRatioBasedSampler
```

Within the `ParentBasedSampler` the most common is the
`TraceIdRatioBasedSampler`. It deterministically samples a percentage of traces

> that you pass in as a parameter.

Reviewer 4: So this came up in the python PR, but I'm not really sure what the right recommendation is.

One gotcha with setting parentbased+traceidratio is that if you don't control an upstream service, your sampling rate isn't actually respected.

And the inverse gotcha is that if you have multiple services all using traceidratio and they set different sampling rates, you'll get turnt too.

Got an inkling towards a preferred setting + guidance? We'd need to be uniform across all languages, so I wanna make sure it's a coherent bit of guidance for everyone.

Author: Hm, yea, maybe needs some guidance around setting up sampling when dealing with a "public" endpoint.

But even in that case I think `ParentBased` with `TraceIdRatioBased` is the way to go, its just that you might want to configure `remote_parent_sampled` to also be `TraceIdRatioBased` instead of `AlwaysOn`.

> ##### Environment Variables

Reviewer: ```suggestion
#### Environment Variables
```

You can configure the `TraceIdRatioBasedSampler` with environment variables:

```shell
export OTEL_TRACES_SAMPLER="parentbased_traceidratio"
export OTEL_TRACES_SAMPLER_ARG="0.1"

> ```

Reviewer 3: Earlier you define:

> - `root` - No parent
> - `remote_parent_sampled` - Parent is from a remote Span that is sampled
> - `remote_parent_not_sampled` - Parent is from a remote Span that is not sampled
> - `local_parent_sampled` - Parent is from a local Span that is sampled
> - `local_parent_not_sampled` - Parent is from a local Span that is not sampled

I see that these are defined lower for Erlang terms. Is there any way to configure these with the ENV vars method?

Author: There isn't. I see one place I can make this clearer, I'll update the paragraph in `ParentBasedSampler` section to be:

> The sampler can be configured with the environment variables `OTEL_TRACES_SAMPLER`
and the sampler to use when there is no parent in `OTEL_TRACES_SAMPLER_ARG`, or using the Application configuration allows you
to configure each of the 5 potential states of a Span's parent:

This tells the SDK to sample spans such that only 10% of Traces get created.

> ##### Application configuration

Reviewer: ```suggestion
#### Application configuration
```

Example in the Application configuration with a root sampler for sampling 10% of
Traces and using the parent decision in the other cases:

<!-- prettier-ignore-start -->
{{< tabpane langEqualsHeader=true >}}

{{< tab Erlang >}}
%% config/sys.config.src

> {sampler, {parent_based, #{root => {trace_id_ratio_based, 0.1},
>                            remote_parent_sampled => always_on,
>                            remote_parent_not_sampled => always_off,
>                            local_parent_sampled => always_on,
>                            local_parent_not_sampled => always_off}}}
> {{< /tab >}}
> 
> {{< tab Elixir >}}
> # config/runtime.exs
> sampler: {:parent_based, %{root: {:trace_id_ratio_based, 0.1},
>                            remote_parent_sampled: :always_on,
>                            remote_parent_not_sampled: :always_off,
>                            local_parent_sampled: :always_on,
>                            local_parent_not_sampled: :always_off}}

Reviewer 3: ```suggestion
{sampler, {parent_based, #{root => {trace_id_ratio_based, 0.10},
                           remote_parent_sampled => always_on,
                           remote_parent_not_sampled => always_off,
                           local_parent_sampled => always_on,
                           local_parent_not_sampled => always_off}}}
{{< /tab >}}

{{< tab Elixir >}}
# config/runtime.exs
sampler: {:parent_based, %{root: {:trace_id_ratio_based, 0.10},
                           remote_parent_sampled: :always_on,
                           remote_parent_not_sampled: :always_off,
                           local_parent_sampled: :always_on,
                           local_parent_not_sampled: :always_off}}
```

Absolutely minor thing, but since you state "10%" in the text, then using `0.10` makes it easier to visually match the content than a `0.1` even if it's absolutely the same value in the end.

{{< /tab >}}

{{< /tabpane >}}
<!-- prettier-ignore-end -->

> #### AlwaysOn and AlwaysOff Sampler

Reviewer: ```suggestion
### AlwaysOn and AlwaysOff Sampler
```

The other two built-in samplers are the `AlwaysOnSampler` and the
`AlwaysOffSampler`.

> ##### Environment Variables

Reviewer: ```suggestion
#### Environment Variables
```

You can configure the `ParentBasedSampler` to use either the `AlwaysOnSampler`
or `AlwaysOffSampler` with the environment variable `OTEL_TRACES_SAMPLER`:

```shell
export OTEL_TRACES_SAMPLER="parentbased_always_on"
```

And for the `AlwaysOffSampler`:

```shell
export OTEL_TRACES_SAMPLER="parentbased_always_off"
```

> ##### Application configuration

Reviewer: ```suggestion
#### Application configuration
```

> Example in the Application configuration with a root sampler that always samples

Reviewer 3: ```suggestion
Here's an example in the Application configuration with a root sampler that always samples
```

and using the parent decision in the other cases:

<!-- prettier-ignore-start -->
{{< tabpane langEqualsHeader=true >}}

{{< tab Erlang >}}
%% config/sys.config.src
{sampler, {parent_based, #{root => always_on,
                           remote_parent_sampled => always_on,
                           remote_parent_not_sampled => always_off,
                           local_parent_sampled => always_on,
                           local_parent_not_sampled => always_off}}}
{{< /tab >}}

{{< tab Elixir >}}
# config/runtime.exs
sampler: {:parent_based, %{root: :always_on,
                           remote_parent_sampled: :always_on,
                           remote_parent_not_sampled: :always_off,
                           local_parent_sampled: :always_on,
                           local_parent_not_sampled: :always_off}}

{{< /tab >}}

{{< /tabpane >}}
<!-- prettier-ignore-end -->

> ### Custom Sampler

Reviewer: ```suggestion
## Custom Sampler
```
